SUGAR - non DAO service end point

minsunny opened the pull request:
#3282

Service Unified GAteway Relay called "SUGAR" is a webAgent which allows to call services not DAOs with its arguments.

Requirements :

Service List
Methods List in the service according to the step 1
Parameters for the method according to the step 2
Providing Parameters Info
Creating dynamic URL
Add modelled Exceptions

minsunny committed:
SUGAR, temp commit for the help request

minsunny committed:
sugar temp commit

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/NP-3282/SUGAR

minsunny committed:
temp commit before testing

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/NP-3282/SUGAR

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/NP-3282/SUGAR

minsunny committed:
Merge branch 'f/NP-3282/SUGAR' of github.com:minsunny/foam2 into f/NP-3282/SUGAR

minsunny committed:
dealt with an exception

minsunny committed:
completed sugar

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/NP-3282/SUGAR

deejes commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,257 @@
+  package: 'foam.nanos.dig',
+  name: 'SUGAR',
+
+  documentation: 'Service Unified GAteway Relay - Perform non-DAO operations against a web service',
```
Case error in gateway

jlhughes replied:
That's intentional as it is the GA in SUGAR.

deejes commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,257 @@
+        this.argumentInfo[j].sub(function(argInfo) {
+          self.flag = !self.flag;
+        });
+      }
```
function logic needs indentation

minsunny replied:
done

deejes commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,257 @@
+        }
+
+        for ( let j = 0 ; j < argumentInfo.length ; j++ ) {
+        var paramUrl = "";
```
indentation for function body required

minsunny replied:
done

jlhughes reviewed (commented):
Please add your tests to this PR.  
Also, if you don't have a test for 'flag', at least add a test with comments regarding what the flag format would be.

jlhughes commented on src/foam/nanos/boot/NSpec.js:
```diff
@@ -113,6 +113,11 @@ foam.CLASS({
       displayWidth: 80
     },
+    {
+      class: 'String',
```
Please add documentation as to the purpose and use of this field.

jlhughes commented on src/foam/nanos/dig/Argument.js:
```diff
@@ -0,0 +1,30 @@
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.dig',
```
Please document the class and properties as to purpose and use.

jlhughes commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -0,0 +1,173 @@
+import java.lang.reflect.*;
+import java.lang.Exception;
+
+public class SugarWebAgent
```
Please document.

kgrgreer commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,257 @@
+
+        var filteredMethod =
+          methods.filter(function(fm) {
+            for ( var j = 0; j < fm.args.length; j++ ) {
```
spaces before ;'s

kgrgreer commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,257 @@
+          }).map(function(m) { return m.name; }).sort();
+
+        if ( filteredMethod.length > 0) {
+          for ( var i = 0; i < methods.length; i++ ) {
```
spaces before ;'s

kgrgreer commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,257 @@
+
+          var filteredMethod =
+            methods.filter(function(fm) {
+              for ( var j = 0; j < fm.args.length; j++ ) {
```
spaces before ;'s

kgrgreer commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,257 @@
+
+          var methods = of.getOwnAxiomsByClass(foam.core.Method);
+
+          for ( var i = 0; i < methods.length; i++ ) {
```
spaces before ;'s

kgrgreer commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,257 @@
+    },
+    {
+      class: 'URL',
+      // TODO: appears not to work if named 'url', find out why.
```
Please open a ticket to track this issue.

kgrgreer commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,257 @@
+            }
+          }
+        } else { // use this flag to give a change event for argumentInfo (FObjectArray)
+          for ( let k = 0 ; k < argumentInfo.length ; k++ ) {
```
Us 'var' instead of 'let' if not trapping value in a closure (function).

kgrgreer commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -0,0 +1,173 @@
+package foam.nanos.dig;
+
+import foam.lib.json.JSONParser;
```
sort imports

kgrgreer commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -0,0 +1,173 @@
+      Class[] paramTypes = null; // for picked Method's parameters' types
+      Object arglist[] = null; // to store each parameters' values
+
+      for ( int k = 0; k < m.length; k++ ) {
```
spaces before ;'s, here and elsewhere

kgrgreer commented on src/foam/nanos/dig/exception/EmptyParameterException.js:
```diff
@@ -0,0 +1,33 @@
+      value: '400'
+    },
+    {
+      class: 'String',
```
Should this be an Int? Same as status?

minsunny committed:
handled the comments

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/NP-3282/SUGAR

shreymahendru commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,256 @@
+
+        if ( ! of ) return;
+
+        this.interfaceName = of.id.toString();
```
You have the logic of these `if` conditions more than one place.... you should abstract that into a function of its own.

shreymahendru commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,256 @@
+        //var methodName = methods.map(function(m) { return m.name; }).sort();
+
+        var filteredMethod =
+          methods.filter(function(fm) {
```
Have a function that abstracts the filtering of methods, that should return a single method since you are only concerned with the 0th element of the filteredMethod. (Correct me if I am wrong)

shreymahendru commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,256 @@
+
+        if ( filteredMethod.length > 0 ) {
+          for ( var i = 0 ; i < methods.length ; i++ ) {
+            if ( methods[i].name == filteredMethod[0] ) {
```
Not sure if we support overriding on services' method, if we do, filtering would be a problem just by name. 
If we don't, why don't you just break the loop once you find the correct method with the same name as filteredmethod?

shreymahendru replied:
`array.find` would work here if you are trying to search a **single** method by name.

shreymahendru commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -0,0 +1,177 @@
+
+      Class c = Class.forName(interfaceName);
+
+      Method m[] = c.getMethods();  // get Methods' List from the class
```
you might want to use more descriptive names for your variables

shreymahendru commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -0,0 +1,177 @@
+
+            logger.debug(pArray[j].getName() + " :   " + p.getParameter(pArray[j].getName()));
+
+            // casting and setting according to parameters type
```
Switch would be perfect and more efficient/readable for this

minsunny replied:
conditional expressions are different depending on primitive types or not though..?

shreymahendru replied:
Sorry. I don't understand... But you can have 
```
String typeName = pArray[j].getType().getCanonicalName() 
switch(typeName) {
case "double": 
arglist[j] = Double.parseDouble(p.getParameter(pArray[j].getName()));
case "boolean":
arglist[j] = Boolean.parseBoolean(p.getParameter(pArray[j].getName()));
....
default:
DigErrorMessage
}
```

kgrgreer replied:
In the future, we're going to switch method arguments to FOAM modelled types, the same as for properties. At that point we'll be able to just use the parser provided by the PropertyInfo, which is better than either switch or many if/elses. Until then, a switch would be better than ifelseifelse's. Another method would be to define a static hashmap of String to Parser mappings and then just lookup the parser to use.

shreymahendru commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -0,0 +1,177 @@
+{
+  public SugarWebAgent() {}
+
+  public void execute(X x) {
```
Please divide the execute function into smaller functions with descriptive names (something like steps), this makes it more readable/ enjoyable for the second dev.

minsunny committed:
handled the comments on PR
 - splited the function on sugarWebAgent
 - change from String to Int on Exceptions codes
 - etc. on PR

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/NP-3282/SUGAR

minsunny committed:
deleted innerscripts in DigWebAgent

minsunny committed:
modified claases

minsunny committed:
Merge branch 'f/NP-3282/SUGAR' of github.com:minsunny/foam2 into f/NP-3282/SUGAR

kgrgreer commented on src/foam/nanos/dig/SUGAR.js:
```diff
@@ -0,0 +1,259 @@
+        if ( ! service.cls_.getAxiomByName('delegate') ) {
+          this.interfaceName = "";
+          return;
+        } else {
```
No need for 'else' block since you've already returned in the 'if'. Just remove the 'else' and move the code in that block two spaces to the left.

kgrgreer commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -0,0 +1,194 @@
+        Method method_[] = class_.getMethods();  // get Methods' List from the class
+
+        for (int k = 0; k < method_.length; k++) {
+          if (method_[k].getName().equals(methodName)) { //found picked Method
```
fix spacing

kgrgreer commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -0,0 +1,194 @@
+            paramTypes = new Class[pArray.length];
+            arglist = new Object[pArray.length];
+
+            for (int j = 0; j < pArray.length; j++) {  // checking the method's each parameter
```
fix spacing

kgrgreer commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -0,0 +1,194 @@
+        }
+      }
+
+      executeMethod(x, resp, out, class_, serviceName, methodName, paramTypes, arglist);
```
Do we want to call this if class_ is null?

kgrgreer commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -0,0 +1,194 @@
+        .build();
+      outputException(x, resp, "JSON", out, error);
+
+      return;
```
No need for return.

kgrgreer commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -0,0 +1,194 @@
+        .build();
+      outputException(x, resp, "JSON", out, error);
+
+      return;
```
No need for return.

minsunny committed:
fixing spacing and extra elses

your changes. Lines starting

minsunny committed:
fixed comflicts

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/NP-3282/SUGAR

minsunny committed:
fixing ui bugs

minsunny committed:
temp commit

minsunny committed:
merge
Merge branch 'master' of github.com:foam-framework/foam2 into f/NP-3282/SUGAR

minsunny committed:
Handling Object Parameters

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/NP-3282/SUGAR

minsunny committed:
changed code for helping optimizing

minsunny committed:
minor modification adding return, checking blank etc.

jlhughes commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -192,4 +178,81 @@ protected void outputException(X x, HttpServletResponse resp, String format, Pri
+              try {
+                fieldList[m].set(clsObj, prop);
+              } catch (IllegalAccessException e) {
+                e.printStackTrace();
```
Can this be logged?

jlhughes commented on src/foam/nanos/dig/SugarWebAgent.java:
```diff
@@ -192,4 +178,81 @@ protected void outputException(X x, HttpServletResponse resp, String format, Pri
+        }
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
```
Can this be logged?

jlhughes reviewed (commented):
Changes look OK. 
Couple comments regarding logging.

minsunny committed:
handling enum parameters

minsunny committed:
changed error message

minsunny committed:
removed return

minsunny committed:
deleted additional else

kgrgreer commented on src/foam/nanos/dig/Argument.js:
```diff
@@ -0,0 +1,58 @@
+      label: 'Object Value',
+      documentation: 'Parameters input value (Object Type Parameters)',
+      factory: function() {
+        var model = this.lookup(this.javaType.toString(), true);
```
javaType is already a String, so you don't need to call toString().

kgrgreer commented on src/foam/nanos/dig/Argument.js:
```diff
@@ -0,0 +1,58 @@
+    },
+    {
+      class: 'FObjectProperty',
+      of: this.javaType,
```
This line can't work since 'this' isn't defined here.

kgrgreer commented on src/foam/nanos/dig/Argument.js:
```diff
@@ -0,0 +1,58 @@
+
+        if ( this.javaType != '' && ( this.javaType != 'String' && this.javaType != 'double' && this.javaType != 'boolean' && this.javaType != 'long' ) )
+          return model.create(null, this);
+        else return null;
```
No need for 'else' since you've already returned.

minsunny committed:
handled comments on PR

minsunny committed:
merge updated
Merge branch 'master' of github.com:foam-framework/foam2 into f/NP-3282/SUGAR

jlhughes reviewed (approved):
Approved, minus Shrey's comments, which are valid, but will have to be dealt with at a later time.
